libraries/eosiolib/contracts/eosio/key_value.hpp

```diff
@@ -0,0 +1,870 @@
+   return key_type{prefix + key};
+}
+
+template <typename I>
+inline I flip_msb(I val) {
+   constexpr static size_t bits = sizeof(I) * 8;
+   return val ^ (static_cast<I>(1) << (bits - 1));
+}
+
+template <typename I>
+inline I get_msb(I val) {
+   constexpr static size_t bits = sizeof(I) * 8;
+   return val >> (bits - 1);
+}
+
+template <typename I, typename std::enable_if_t<std::is_integral<I>::value, int> = 0>
+inline key_type make_key(I val) {
+   if (std::is_signed<I>::value) {
+      val = flip_msb(val);
+   }
+
+   auto big_endian = swap_endian<I>(val);
+
+   const char* bytes = reinterpret_cast<char*>(&big_endian);
+   constexpr size_t size = sizeof(big_endian);
+   key_type s(bytes, size);
+   return s;
+}
+
+template <typename I, typename F>
+inline key_type make_floating_key(F val) {
+   if (val == -0) {
+      val = +0;
+   }
+
+   auto* ival = reinterpret_cast<I*>(&val);
+   I bit_val;
+   auto msb = get_msb(*ival);
+   if (msb) {
+      // invert all the bits
+      bit_val = ~(*ival);
+   } else {
+      // invert just msb
+      bit_val = flip_msb(*ival);
+   }
+
+   auto big_endian = swap_endian<I>(bit_val);
+
+   const char* bytes = reinterpret_cast<char*>(&big_endian);
+   constexpr size_t size = sizeof(big_endian);
+   key_type s(bytes, size);
+   return s;
+}
+
+inline key_type make_key(float val) {
+   return make_floating_key<uint32_t>(val);
+}
+
+inline key_type make_key(double val) {
+   return make_floating_key<uint64_t>(val);
+}
+
+inline key_type make_key(const char* str, size_t size, bool case_insensitive=false) {
+   size_t data_size = size + 3;
+   void* data_buffer = data_size > detail::max_stack_buffer_size ? malloc(data_size) : alloca(data_size);
+
+   if (case_insensitive) {
+      std::transform(str, str + size, (char*)data_buffer, [](unsigned char c) -> unsigned char { return std::toupper(c); });
+   } else {
+      memcpy(data_buffer, str, size);
+   }
+
+   ((char*)data_buffer)[data_size - 3] = 0x01;
+   ((char*)data_buffer)[data_size - 2] = 0x00;
+   ((char*)data_buffer)[data_size - 1] = 0x00;
+
+   key_type s((const char*)data_buffer, data_size);
+
+   if (data_size > detail::max_stack_buffer_size) {
+      free(data_buffer);
+   }
+   return s;
+}
+
+inline key_type make_key(const std::string& val, bool case_insensitive=false) {
+   return make_key(val.data(), val.size(), case_insensitive);
+}
+
+inline key_type make_key(const char* str, bool case_insensitive=false) {
+   return make_key(std::string{str}, case_insensitive);
+}
+
+inline key_type make_key(eosio::name n) {
+   return make_key(n.value);
+}
+
+inline key_type make_key(key_type&& val) {
+   return val;
+}
+
+template <typename S, typename std::enable_if_t<std::is_class<S>::value, int> = 0>
+inline key_type make_key(S val) {
+   size_t data_size = 0;
+   size_t pos = 0;
+   void* data_buffer;
+
+   boost::pfr::for_each_field(val, [&](auto& field) {
+      data_size += sizeof(field);
```

Comment:
Doesn't work for most member types: string, vector, ...

Reply:
This will be addressed once the abieos refactor is complete.

Reply:
see #850